### ISSUE TITLE
Render pass compatibility - attachment indices underspecified?

### ISSUE THREAD

**Issue by schwa423** [KhronosGroup/Vulkan-Docs#403]
According to the wording of section 7.2 of 1.0.32 spec, I believe that the following two render passes would be deemed compatible, but should not be:

1) When creating the first render pass, the array of VkAttachmentDescriptions has a description of the sole color attachment at index 0, and the sole depth attachment at index 1.  The sole subpass has a VkAttachmentReference corresponding to each of these.

2) When creating the second render pass, the VkAttachmentDescriptions for color and depth attachments are identical, except that they appear in the opposite order (color at index 1, and depth at index 0).  Again, the sole subpass has a VkAttachmentReference for each of these (with indices swapped, of course).

If these two render passes are compatible, then a Framebuffer created with attachments matching render pass 1 would have problems when rendered with render pass 2, because the attachments would be in the wrong order.  Unless I'm missing something, this seems obviously wrong.

Perhaps the wording:

`Two attachment references are compatible if they have matching format and sample count... `

... should read something like:

`Two attachment references are compatible if they have the same index, and (if the index is not VK_ATTACHMENT_UNUSED) the referenced attachments have matching format and sample count... `

**Comment by krOoze** [comment]
In your counter-example they would not have identical `format`, would they?

**Comment by schwa423** [comment]
Regarding my counter example, I'm not sure what you mean.  Let's say that the `VkFramebuffer framebufferCreatedWithRenderPass1` is created using `renderPass1`.  The attachments in the `VkFramebufferCreateInfo` would need to match those specified in `renderPass1` (i.e. the first attachment is color, and the second is depth).  If you tried to create another framebuffer using the same `VkFramebufferCreateInfo` with the only difference being that `renderPass2` is used instead, this should fail because the formats of the attachments don't match.

However, according to the wording in the "Render Pass Compatibility" section, it seems that the following is legal:

```C++
vk::RenderPassBeginInfo info;
info.renderPass = renderPass2;
info.framebuffer = framebufferCreatedWithRenderPass1;
// etc.
```

If this is intentional, then the Vulkan implementation would need to notice that the formats don't match, and "remap" them.  This is simple in the case where there is one color and one depth attachment, but (I think) ambiguous in general.  For example, what if there are two color attachments and one depth attachment?  Say that `renderPass1` is created with attachments ordered "color1 color2 depth" and `renderPass2` with the order "depth color1 color2".  Assuming that "color1" and "color2" have the same format, how does the Vulkan implementation decide which attachments "correspond"?  Assume that the program's intent is for "color1" to map to "color1", and "color2" to map to "color2".  The spec doesn't guarantee this... an implementation might notice that the attachments at index 1 ("color1" in `renderPass1` and "color2" in `renderPass2`) have compatible formats, and not "remap" them.  

In other words, if this is intended, it seems that the spec must specify how implementations should perform the remapping.

If this is unintended, then the wording in the "Render Pass Compatibility" section should make this clear without requiring the reader to jump to another (unmentioned) section of the spec. 

Also, if this is unintended, then your rhetorical question is right on target:
> what's the point of having Render Pass Compatibility mechanism in the first place?
It would be useful for the spec to contain examples of intended usage, as is often done in OpenGL extensions.

Hopefully that made sense.  I can elaborate if necessary.

**Comment by schwa423** [comment]
Regarding:
> The "identical" makes me uneasy considering what is said in the previous spec paragraphs (seems to override them).

I completely agree.

To grok this section myself, I rewrote it for myself as follows (not sure if it captures the intent of the spec authors):

- 2 RenderPasses are compatible if each corresponding pair of subpasses are compatible
- 2 subpasses are compatible if each corresponding pair of AttachmentReferences are compatible
- 2 AttachmentReferences are compatible if
  - both is either nullptr or UNUSED, or neither are and:
  - both have the same format and sample count

To me, this is very linear and much easier to understand (but again, might not match the intent of the spec authors).

**Comment by krOoze** [comment]
> Regarding my counter example, I'm not sure what you mean.

I tend to edit my posts too much and I am not sure which version you seen. I realized it's not clear and clarified.

What I mean is this VU section requirement of `VkFramebufferCreateInfo`:

> - Any given element of `pAttachments` **must** have been created with an `VkFormat` value that matches the `VkFormat` specified by the corresponding `VkAttachmentDescription` in `renderPass`

So the example would be invalid Vulkan either way.
One `VkImageView` would be some color format and the other depth format. If you have them swapped in `VkRenderPass` this requirement is violated.

**Comment by janharaldfredriksen-arm** [comment]
@schwa423:
I'm not clear on whether you're talking about the ordering of the Attachment Descriptions (VkRenderPassCreateInfo::pAttachments) or the Attachment References in the VkSubpassDescription structure?

Assuming attachment references, I think you're missing the second condition in your initial example:
>Two arrays of attachment references are compatible if all corresponding pairs of attachments are compatible.

"corresponding pairs" here means what I think you're trying to say with the "have the same index" language. The implementation is not supposed to do any remapping of attachment references.

Re: the "identical" language:
>If two render passes contain more than one subpass, they are compatible if they are identical ..

This could perhaps be made clearer. It is a much stronger statement than the previous paragraph describing the single subpass case. Remember that implementations may merge subpasses. The (implementation-specific) algorithm for this is likely to depend on the subpass descriptions and subpass dependencies. Therefore these descriptions must: match for two render passes with two or more subpasses to be compatible. The attachment references must also match - just like for the single subpass case, which leaves the few things explicitly listed as the only things that can change between compatible render passes.

**Comment by schwa423** [comment]
@janharaldfredriksen-arm:
I think I'm talking about both the ordering of Attachment Descriptions and Attachment References, but in different contexts :-)

I think that you may have just triggered some understanding in my mind.  Let me first clarify the following... am I correct to say:

For the purposes of creating a Framebuffer, compatibility with a RenderPass is according to the Attachment Descriptions (not the References, because there are multiple arrays of References) that the RenderPass was created with.

(if I'm wrong about the above, then I'm very confused!)

Regarding:
> Assuming attachment references, I think you're missing the second condition in your initial example:
>> Two arrays of attachment references are compatible if all corresponding pairs of attachments are compatible.

>"corresponding pairs" here means what I think you're trying to say with the "have the same index" language. The implementation is not supposed to do any remapping of attachment references.
... the term "corresponding pairs" is ambiguous (at least taken at face value... it may be possible to arrive at an unambiguous interpretation by applying the other constraints implied by other wording).  It might mean two things:
- that the indices of the Attachment References might differ, as long as the referred-to Attachment Descriptions are compatible
- the indices of the Attachment References must be the same, AND the referred-to Attachment Descriptions must be compatible.

When I said "have the same index", I meant the second interpretation.  However, I'm more uncertain than ever about which interpretation is intended.  Can you shed any light on this?

In particular, if the first interpretation is intended, then the problem remains with using `framebufferCreatedWithRenderPass1` with `renderPass2` (that's why I came up with the second interpretation).  

I have thoughts about:
> Re: the "identical" language:

... but I'll leave them until the above is cleared up.

**Comment by schwa423** [comment]
(oops, updated previous comment... I accidentally included some new text that I wrote as part of a quote attributed to @janharaldfredriksen-arm)

**Comment by janharaldfredriksen-arm** [comment]
>For the purposes of creating a Framebuffer, compatibility with a RenderPass is according to the Attachment Descriptions (not the References, because there are multiple arrays of References) that the RenderPass was created with.

Yeah, I understand what you're saying, but I think the confusion (including mine) comes about because what you're calling "compatibility with a RenderPass" in that sentence is not the same thing as what's described in the Render Pass Compatibility section.

The valid usage for VkFramebufferCreateInfo is specified in terms of the VkAttachmentDescriptions. So in (a slightly extended version of) your example:
RenderPass1 has:
- pAttachments = [color, depth]
- Subpass 0: pColorAttachments[0].attachment = 0, pDepthStencilAttachment[0].attachment = 1

RenderPass2 has:
- VkAttachmentDescription[depth, color]
- Subpass 0: pColorAttachments[0].attachment = 1, pDepthStencilAttachment[1].attachment = 0

If the VkFramebufferCreateInfo has pAttachments[color, depth], then only RenderPass1 can be used to create a Framebuffer; RenderPass2 runs foul of the VU section that @krOoze pointed out.

Render pass compatibility has no bearing on this - only the VU statement.

Now, when you try to use the Framebuffer, say by passing it to vkCmdBeginRenderPass, then (only) the Render pass compatibility matters, and this is defined in terms of Attachment References. For color attachments, the check for compatible formats is:
`foreach(color attachment i)
{
    format1 = RenderPass1.pAttachments[pColorAttachments[0].attachment].format;
    format2 = RenderPass2.pAttachments[pColorAttachments[0].attachment].format;
    is_compatible = (format1 == format2);
}`

.. and similar for the other fields. (You can see how the validation layers do the complete check for this at https://gitlab.khronos.org/vulkan/LoaderAndValidationLayers/blob/1.0/layers/core_validation.cpp#L2352)

This would make these render passes compatible.

**Comment by schwa423** [comment]
Thanks for pointing me at the validation layer code; I should have thought of that.  I now have a crystal-clear understanding of what the spec authors mean by the term "Render Pass Compatibility".

Incidentally, the validation algorithm is very similar to my intent when I earlier wrote: 

> To grok this section myself, I rewrote it for myself as follows (not sure if it captures the intent of the spec authors):
- 2 RenderPasses are compatible if each corresponding pair of subpasses are compatible
- 2 subpasses are compatible if each corresponding pair of AttachmentReferences are compatible
- 2 AttachmentReferences are compatible if
  - both is either nullptr or UNUSED, or neither are and:
  - both have the same format and sample count

I find the validation code to be much easier to understand than the spec.  

One reason, as already mentioned above, is the confusing language about render passes with single subpasses vs. multiple subpasses.  The code shows that nothing special is happening in the single-subpass case.   However, my natural assumption (probably the same assumption that many hackers would make) is that the single-subpass case wouldn't be mentioned separately unless some specific nuance couldn't be captured by the multi-subpass language.

Is there any chance that pseudo-code for the validation algorithm could be included in the spec, instead of the current language?

**Comment by janharaldfredriksen-arm** [comment]
>Is there any chance that pseudo-code for the validation algorithm could be included in the spec, instead of the current language?

I'll see what I can do to make it clearer at least. (But it will be a few days before I get to it.)

(Note to self: it's not clear if the "identical" language intentionally requires the attachment descriptions to also match.)

**Comment by schwa423** [comment]
(commenting separately, because this is orthogonal to the "Render Pass Compatibility" wording).

Now that we have the same understanding of "Render Pass Compatibility", I still don't understand whether two compatible RenderPasses are guaranteed to be usable with Framebuffer created using one of them, and if so, how this is accomplished.

Here's a scenario:

RenderPass1 has:
- pAttachments = [depth, color]
- Subpass 0: pColorAttachments[0].attachment = 1, pDepthStencilAttachment[0].attachment = 0

RenderPass2 has:
- pAttachments = [depth, color1, color2]
- Subpass 0: pColorAttachments[0].attachment = 2, pDepthStencilAttachment[1].attachment = 0

I think we agree that these are compatible, according to "Render Pass Compatibility".

Next, we create a Framebuffer using RenderPass1:
- renderPass = RenderPass1
- pAttachments = [depthImageView, colorImageView].

Clearly we can use Framebuffer with RenderPass1.  But can we use it with RenderPass2?

(this is the same issue I was getting at earlier, but without the confusion about what we mean by "Render Pass Compatibility")

According to the validation code, it looks like Framebuffer is incompatible with RenderPass2.  For example, in `VerifyFramebufferAndRenderPassLayouts()` I see the following code:

```C++
   if (pRenderPassInfo->attachmentCount != framebufferInfo.attachmentCount) {
        skip_call |= log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, (VkDebugReportObjectTypeEXT)0, 0, __LINE__,
                             DRAWSTATE_INVALID_RENDERPASS, "DS", "You cannot start a render pass using a framebuffer "
                                                                 "with a different number of attachments.");
    }
```

However, in Section 7.3 in the discussion of `VkFramebufferCreateInfo`, the `renderPass` field is described as:
> a render pass that defines what render passes the framebuffer will be compatible with. See Render Pass Compatibility for details.

This seems to imply that a Framebuffer can be used as the target of any RenderPass that is compatible (as defined in "Render Pass Compatibility") with the RenderPass that was used to create the Framebuffer.  This is contradicted by the validation code above.

If we agree that the wording in Section 7.2 is correct (albeit not optimally clear), then it seems like the wording in Section 7.3 is misleading at best.

This is the fundamental confusion that caused me to file this issue; now that we now using the same terminology, we can begin to grapple with it.

**Comment by schwa423** [comment]
Splitting off a separate issue #404, so that @janharaldfredriksen-arm can close this one when he has a chance to update the wording in section 7.2

**Comment by janharaldfredriksen-arm** [comment]
Section 7.2 got updated a few revision. (I don't remember exactly when, but it's at least included in 1.0.43.)

I ended up just collapsing the single and multi sub-pass cases, which I think gets rid of the ambiguity here.

**Comment by CarloWood** [comment]
This issue is not resolved. I got here because I couldn't understand the wording of the current spec (version 1.2).

Reading this thread the following happened:
* Every post by schwa423 is crystal clear (and exactly describing my issue with understanding the spec).
* Posts by others are not clear at all to me; they gave me the impression that schwa423's posts were not understood.
* Also I *only* finally got an answer to my question by looking at the validation code (not from any post here).

My endeavour started with the following post on the vulkan discord server:

Can two render passes that have indices that are not exchangeable still be compatible?
From spec: Two attachment references are compatible if they have matching format and sample count, or are both VK_ATTACHMENT_UNUSED or the pointer that would contain the reference is NULL.
That is not clear at all.

Immediately after I found this issue and starting reading that. It is now 1 hour later.

What is not clear about this sentence is that, for a start, an 'attachment reference' does not contain a format or sample count.
nor does it contain a pointer.

From "8.2. Render Pass Compatibility" :

* Two render passes are compatible if their corresponding color, input, resolve, and depth/stencil attachment references are compatible and if they are otherwise identical except for [...]

That says that for each index {0, 1, 2, ...} into the SubpassDescription arrays pColorAttachments, pInputAttachments, pResolveAttachments, and the (non-array) pDepthStencilAttachment the "vk::AttachmentReference" must be "compatible".

Now, a render pass --what this is about-- is created with a RenderPassCreateInfo that has possibly multiple SubpassDescription's. Namely the pSubpasses parameter.

So, apparently compatibility must exist for each pair of indices (i, j) where i is the index into the pSubpasses array and j is the index into an array of SubpassDescription - and that for each of the existing arrays.

Most notably, it is not required that the 'attachment' parameter of the AttachmentReference2 struct in those arrays match, but only the AttachmentDescription structs that they refer to through their corresponding pAttachments array of the same RenderPassCreateInfo used to create the renderpass.

But the spec doesn't talk about attachment descriptions, it talks about attachment references.

I think that exactly that what schwa423 describes here has still not been addressed.

**Comment by CarloWood** [comment]
I formulated my own view on the matter. Correct me if I am wrong (copied from discord again):

From what I understand so far, there are four types of indices involved:
1) subpass indices - an index into VkRenderPassCreateInfo::pSubpasses. Since this is an array that is part of VkRenderPassCreateInfo the index is specific to a giving renderpass instance. It is not clear if two compatible renderpasses must have identical subpasses and/or their subpass indices need to be somehow interchangeable.
2) For a given VkSubpassDescription (one element of VkRenderPassCreateInfo::pSubpasses) there are two indices: an index into pInputAttachments, and an index into pColorAttachments. If pResolveAttachments is non-NULL than that array uses the same index (type) as pColorAttachments. The uint32_t's in the pPreserveAttachments are of the same type as those of the VkAttachmentReference::attachment elements of the VkAttachmentReferences in each of the arrays (and *pDepthStencilAttachment).
3) The VkRenderPassCreateInfo::pAttachments arrays, whose index is used in VkAttachmentReference::attachment.

Each 'type' of index corresponds with both a size and an ordering (of course).
Lets call these index types SubpassIndex, InputAttachmentIndex, ColorAttachmentIndex and RenderpassAttachmentIndex.
Then any given VkAttachmentReference, whose attachment index of the type RenderpassAttachmentIndex is also renderpass specific (the same renderpass as the RenderpassAttachmentIndex is bound to).
However, the arrays that contain VkAttachmentReferences, namely pInputAttachments and pColorAttachments (pResolveAttachments) are not necessarily renderpass specific, even though their elements are. That is, if the size of these array is the same then two different renderpasses can still have interchangeable InputAttachmentIndex and/or ColorAttachmentIndex, as long as the reference they contain point to a "compatible" VkAttachmentDescription (which means the format and samples are the same).

The spec says, in 8.2 Render Pass Compatibility:
Two render passes are compatible if their corresponding color, input, resolve, and depth/stencil attachment references are compatible and if they are otherwise identical except for:

* Initial and final image layout in attachment descriptions
* Load and store operations in attachment descriptions
* Image layout in attachment references
which simply rules out everything except the format and samples parameter of the VkAttachmentDescription (and in the case of a VkAttachmentReference, the renderpass specific attachment element must point to this VkAttachmentDescription (in the renderpass' pAttachments array (that it was created with)).

Here corresponding means: the same SubpassIndex and within that the same ColorAttachmentIndex and/or InputAttachmentIndex. Note that the requirement that the depth/stencil attachment is compatible is dropped if the renderpasses have only a single subpass.

Furthermore, 8.2 states:
Two arrays of attachment references are compatible if all corresponding pairs of attachments are compatible. If the arrays are of different lengths, attachment references not present in the smaller array are treated as VK_ATTACHMENT_UNUSED.

(Imho this reveals a blatant error in the previous quote - it should have said "attachment reference arrays").

Finally, it states:
Two attachment references are compatible if they have matching format and sample count, or are both VK_ATTACHMENT_UNUSED or the pointer that would contain the reference is NULL.

where, apparently, "have matching..." means "the VkAttachmentDescription that they point to have matching...".
So, since for renderpass compatibility you always go through the VkSubpassDescription arrays, the dependency on (or uniqueness of; or renderpass specificness of) the RenderpassAttachmentIndex drops out.

But the renderpass specificness of SubpassIndex, InputAttachmentIndex and ColorAttachmentIndex do not - with the exception that it is allowed for two compatible renderpasses to have arrays of different length - but extending them with virtual VkAttachmentReference-s objects with a VK_ATTACHMENT_UNUSED value for attachment. And the existing element has an actual value of VK_ATTACHMENT_UNUSED (or the pointer that would contain the reference is NULL; but I have no idea what that refers to).

**Comment by CarloWood** [comment]
Issue #404 refers to the fact that a VkFramebuffer is created with an array, VkFramebufferCreateInfo::pAttachments, that uses an index of type RenderpassAttachmentIndex. It is therefore 100% renderpass specific and can NOT be used with a "compatible" renderpass (that doesn't even take the RenderpassAttachmentIndex into account).

Imho, this is a bug in the spec - still - in 2021.
And also #404 has not been addressed or solved.

**Comment by yserrr** [comment]
Hi,

I'm a CS student contributing to Mesa's V3DV Vulkan driver. I'm currently working on an optimization for framebuffer-local self-dependencies, where input attachment reads can stay in the tile buffer instead of forcing a store/load.

I am currently blocked on the intended semantics of render pass compatibility, specifically whether attachment indices and cross-subpass attachment identity are part of the compatibility contract.

My current reading of the specification is that corresponding color, input, resolve, and depth/stencil attachment references are compared through attachment-reference compatibility, which is based on the referenced attachment format and sample count (or both being `VK_ATTACHMENT_UNUSED`).

As a result, I am not sure whether the following two render passes are intended to be compatible.

Assume attachments 0 and 1 have the same format and sample count in both render passes.

RenderPass A:

Subpass 0:
    color[0] -> attachment 0

Subpass 1:
    input[0] -> attachment 0


RenderPass B:

Subpass 0:
    color[0] -> attachment 0

Subpass 1:
    input[0] -> attachment 1

The corresponding color references are compatible, and the corresponding input references are also compatible by format/sample count.

However, the logical relation is different:

A: subpass 0 output == subpass 1 input
B: subpass 0 output != subpass 1 input

So the actual framebuffer attachment used by the input attachment can differ even though the render passes appear compatible.

VVL also appears to implement compatibility based on the attachment-reference compatibility rules rather than requiring attachment indices to match.

For a driver this distinction matters. In V3DV, for example, an input attachment may be lowered to a direct tile-buffer read. If attachment identity is not preserved across compatible render passes, the driver cannot treat the attachment index from the render pass used at pipeline creation as an invariant. It would instead need to resolve or normalize the attachment relation against the render pass used for the actual render pass instance.

So I would like to clarify the intended contract:

1. Is it intentional that compatible render passes may change the attachment identity relation between subpasses, as in the example above?

2. If so, is the expectation that implementations must treat attachment indices as non-invariant across compatible render passes and map the pipeline's attachment semantics to the actual render pass at execution time?

3. If this example is not intended to be compatible, which part of the compatibility rules is intended to reject it? In particular, does "otherwise identical" imply preservation of attachment indices or of the equivalence relation between attachment references?

I'm mainly trying to understand the intended driver responsibility here.

Thanks!

**Comment by yserrr** [comment]
Resolved: the Vulkan Working Group confirmed that the example above is not intended to be compatible. 

The relevant requirement is that the attachment relationship/mapping must be preserved; attachment-reference compatibility based only on format and sample count is not sufficient when reordering changes that relationship.

This exposed a validation gap rather than requiring a new VU. VVL is being updated to catch this case under VUID-vkCmdExecuteCommands-pBeginInfo-06020 and provide a diagnostic explaining that the attachment references cannot be reordered in this way.

Thanks to everyone who helped clarify this.
